Use more smart pointers for dynamicDowncast<>in SVG code

rwlbuis opened the pull request:
#### 7196d1f760d62c84bcddef4c342d42398a9048bd
<pre>
Use more smart pointers for dynamicDowncast&lt;&gt;in SVG code
<a href="https://bugs.webkit.org/show_bug.cgi?id=272817">https://bugs.webkit.org/show_bug.cgi?id=272817</a>

Reviewed by Chris Dumez.

Use more smart pointers for dynamicDowncast&lt;&gt;in SVG code.

* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::findPreviousAndNextAttributes):
* Source/WebCore/rendering/svg/RenderSVGTransformableContainer.cpp:
(WebCore::RenderSVGTransformableContainer::additionalContainerTranslation const):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleRootOrContainer const):
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
(WebCore::SVGContainerLayout::layoutChildren):
(WebCore::SVGContainerLayout::positionChildrenRelativeToContainer):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::checkForSVGRepaintDuringLayout):
(WebCore::updateObjectBoundingBox):
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
(WebCore::SVGRenderSupport::computeContainerStrokeBoundingBox):
(WebCore::layoutSizeOfNearestViewportChanged):
(WebCore::SVGRenderSupport::layoutChildren):
(WebCore::isPointInCSSClippingArea):
(WebCore::SVGRenderSupport::clipContextToCSSClippingArea):
(WebCore::SVGRenderSupport::calculateApproximateStrokeBoundingBox):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGPaintingFeatures):
(WebCore::writeResources):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::walkTree):
* Source/WebCore/rendering/svg/SVGTextQuery.cpp:
(WebCore::flowBoxForRenderer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::getElementCTM):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::pathOnlyClipping):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGTransformableContainer.cpp:
(WebCore::LegacyRenderSVGTransformableContainer::calculateLocalTransform):
* Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
(WebCore::SVGFEComponentTransferElement::setFilterEffectAttributeFromChild):
* Source/WebCore/svg/SVGFontFaceSrcElement.cpp:
(WebCore::SVGFontFaceSrcElement::createSrcValue const):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::invalidateResourceImageBuffersIfNeeded):

Canonical link: <a href="https://commits.webkit.org/277661@main">https://commits.webkit.org/277661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7cd1ccd1c96e00c340580e2c26a45ad5ce2842fc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48260 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27472 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51214 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50948 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44325 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50565 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33406 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24992 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48842 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25166 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20568 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22646 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6316 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44604 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43328 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52852 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19657 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46757 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24572 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41876 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45676 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10643 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25377 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24295 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -99,7 +99,7 @@ const RenderSVGText* RenderSVGText::locateRenderSVGTextAncestor(const RenderObje
 {
     for (RenderObject* descendant = text; descendant; descendant = descendant->nextInPreOrder(text)) {
-        if (auto* svgInline = dynamicDowncast<RenderSVGInlineText>(*descendant))
+        if (CheckedPtr svgInline = dynamicDowncast<RenderSVGInlineText>(*descendant))
```
Not strictly needed since `layoutAttributes()` is a trivial inline function. If this is hot code, we could consider not making this change.

rwlbuis replied:
Ah, sorry, did not check layoutAttributes properly.

cdumez commented on Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
```diff
@@ -207,10 +209,10 @@ bool SVGContainerLayout::layoutSizeOfNearestViewportChanged() const
 
     ASSERT(ancestor);
-    if (auto* viewportContainer = dynamicDowncast<RenderSVGViewportContainer>(ancestor))
+    if (CheckedPtr viewportContainer = dynamicDowncast<RenderSVGViewportContainer>(ancestor))
```
Not strictly needed, `isLayoutSizeChanged()` is a trivial inline getter.

rwlbuis replied:
Fixed.

cdumez commented on Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
```diff
@@ -207,10 +209,10 @@ bool SVGContainerLayout::layoutSizeOfNearestViewportChanged() const
         return viewportContainer->isLayoutSizeChanged();
 
-    if (auto* svgRoot = dynamicDowncast<RenderSVGRoot>(ancestor))
+    if (CheckedPtr svgRoot = dynamicDowncast<RenderSVGRoot>(ancestor))
```
ditto.

rwlbuis replied:
Fixed.

cdumez commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -93,7 +93,7 @@ const RenderElement& SVGRenderSupport::localToParentTransform(const RenderElemen
     // At the SVG/HTML boundary (aka LegacyRenderSVGRoot), we apply the localToBorderBoxTransform
     // to map an element from SVG viewport coordinates to CSS box coordinates.
-    if (auto* svgRoot = dynamicDowncast<LegacyRenderSVGRoot>(parent))
+    if (CheckedPtr svgRoot = dynamicDowncast<LegacyRenderSVGRoot>(parent))
```
ditto.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
```diff
@@ -269,7 +268,7 @@ void writeSVGPaintingFeatures(TextStream& ts, const RenderElement& renderer, Opt
 {
     if (behavior.contains(RenderAsTextFlag::ShowSVGGeometry)) {
-        if (auto* box = dynamicDowncast<RenderBox>(renderer))
+        if (CheckedPtr box = dynamicDowncast<RenderBox>(renderer))
```
ditto.

rwlbuis replied:
Done.

cdumez commented on Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
```diff
@@ -284,28 +283,28 @@ void writeSVGGraphicsElement(TextStream& ts, const SVGGraphicsElement& svgElemen
     SVGLengthContext lengthContext(&svgElement);
 
-    if (auto* element = dynamicDowncast<SVGRectElement>(svgElement)) {
+    if (RefPtr element = dynamicDowncast<SVGRectElement>(svgElement)) {
```
Not strictly needed since the call site is supposed to have a Ref/RefPtr to svgElement on the stack. It's not this function's job to keep parameters alive.

Same comment applies elsewhere in this function.

rwlbuis replied:
Got it, fixed.

cdumez commented on Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
```diff
@@ -102,7 +102,7 @@ void SVGRenderingContext::prepareToRenderSVGContent(RenderElement& renderer, Pai
 
     if (style.hasPositionedMask()) {
-        if (auto* graphicsElement = dynamicDowncast<SVGGraphicsElement>(*renderer.element()))
+        if (RefPtr graphicsElement = dynamicDowncast<SVGGraphicsElement>(*renderer.element()))
```
Not needed.

rwlbuis replied:
Removed this part.

cdumez commented on Source/WebCore/svg/SVGLinearGradientElement.cpp:
```diff
@@ -151,7 +151,7 @@ bool SVGLinearGradientElement::collectGradientAttributes(LinearGradientAttribute
         auto target = SVGURIReference::targetElementFromIRIString(current->href(), treeScopeForSVGReferences());
-        if (auto* gradientElement = dynamicDowncast<SVGGradientElement>(target.element.get())) {
+        if (RefPtr gradientElement = dynamicDowncast<SVGGradientElement>(target.element.get())) {
             current = *gradientElement;
```
`gradientElement.releaseNonNull()` to avoid refcounting churn. Or keep it as a raw pointer. There was nothing wrong with this code.

rwlbuis replied:
Kept as-is.

cdumez reviewed (approved):
We need to be careful when adopting smart pointers to actually follow the guidelines. Otherwise, we'll make the performance worse without actually improving security.

If you find it difficult, maybe we should wait for the static analyzer bot to be ready.

rwlbuis committed:
Use more smart pointers for dynamicDowncast<>in SVG code
https://bugs.webkit.org/show_bug.cgi?id=272817

Reviewed by Chris Dumez.

Use more smart pointers for dynamicDowncast<>in SVG code.

* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::findPreviousAndNextAttributes):
* Source/WebCore/rendering/svg/RenderSVGTransformableContainer.cpp:
(WebCore::RenderSVGTransformableContainer::additionalContainerTranslation const):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleRootOrContainer const):
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
(WebCore::SVGContainerLayout::layoutChildren):
(WebCore::SVGContainerLayout::positionChildrenRelativeToContainer):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::checkForSVGRepaintDuringLayout):
(WebCore::updateObjectBoundingBox):
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
(WebCore::SVGRenderSupport::computeContainerStrokeBoundingBox):
(WebCore::layoutSizeOfNearestViewportChanged):
(WebCore::SVGRenderSupport::layoutChildren):
(WebCore::isPointInCSSClippingArea):
(WebCore::SVGRenderSupport::clipContextToCSSClippingArea):
(WebCore::SVGRenderSupport::calculateApproximateStrokeBoundingBox):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGPaintingFeatures):
(WebCore::writeResources):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::walkTree):
* Source/WebCore/rendering/svg/SVGTextQuery.cpp:
(WebCore::flowBoxForRenderer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::getElementCTM):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::pathOnlyClipping):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGTransformableContainer.cpp:
(WebCore::LegacyRenderSVGTransformableContainer::calculateLocalTransform):
* Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
(WebCore::SVGFEComponentTransferElement::setFilterEffectAttributeFromChild):
* Source/WebCore/svg/SVGFontFaceSrcElement.cpp:
(WebCore::SVGFontFaceSrcElement::createSrcValue const):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::invalidateResourceImageBuffersIfNeeded):

Canonical link: https://commits.webkit.org/277661@main
